### PR TITLE
Add command `tsort`

### LINKED ISSUE #2982: tsort
Implement a variant of `tsort` in the `exp` package. It will follow the [POSIX standard](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/tsort.html) except for the "ENVIRONMENT VARIABLES" section.

Commands will have the following format:

```
tsort [file]
```

It will do its best to gracefully handle cycles in the input data, but no guarantees are made and warnings may be printed if one or more cycles are detected.

### PR DESCRIPTION
**Pull request by jbduncan**
Add `cmds/exp/tsort`, which aims to implement [POSIX tsort](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/tsort.html) (except for "Environment Variables").

Currently I'm undecided what to do about cycles, as handling them is tricky and POSIX takes no stance on them.

There are still some remaining TODOs in the source code.

Resolves #2982.

### CODE REVIEW

File: cmds/exp/tsort/tsort.go
Comment (rminnich) [inline cmds/exp/tsort/tsort.go:10]: I"m a bit concerned about this. You're copying the whole file in? Can't you pull it in a bit at a time?

**Review by rminnich — COMMENTED** [review]
Did you need to vendor all that code, now that we use modules?

File: cmds/exp/tsort/tsort.go
Comment (jbduncan) [reply cmds/exp/tsort/tsort.go:10]: Good catch. I could probably read one line at a time. WDYT?

**Comment by jbduncan** [comment]
> Did you need to vendor all that code, now that we use modules?

@rminnich Good question! I assumed that all dependencies were vendored in this project. Am I mistaken?

Anyway, my test needs the `cmpopts` package from `github.com/google/go-cmp`, and the build fails if I don't vendor it, so I was wondering what's the recommended way of importing it?

File: cmds/exp/tsort/tsort.go
Comment (rminnich) [reply cmds/exp/tsort/tsort.go:10]: I think that's better. We don't want to give up on systems with "small" -- 64 MiB is small, who knew? -- memory :-)

**Comment by rminnich** [comment]
did you do a go mod update? That usually does it for me?

**Comment by jbduncan** [comment]
> did you do a go mod update? That usually does it for me?

I did, and apparently it doesn't work.

After reverting the vendor update with [48710d2](https://github.com/u-root/u-root/pull/2984/commits/48710d2355234cf737991a3d7995c080a7f9436b), running `go get -u -t ./... && go mod tidy && go mod download`, and finally running `go build ./...`, this is what I get.

This suggests to me that vendoring is mandatory. Furthermore, my IDE doesn't see the `github.com/google/go-cmp/cmp/cmpopts` package.

Am I missing something obvious?

```
go: inconsistent vendoring in /Users/jonathan.bluett-dunc/dev/u-root:
        github.com/ProtonMail/go-crypto@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/beevik/ntp@v1.4.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/cenkalti/backoff/v4@v4.3.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/gliderlabs/ssh@v0.3.7: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/gojuno/minimock/v3@v3.3.7: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/google/go-cmp@v0.6.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/google/go-tpm@v0.9.1-0.20240411180339-1fb84445f623: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/google/uuid@v1.6.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/hugelgupf/vmtest@v0.0.0-20240307030256-5d9f3d34a58d: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/insomniacslk/dhcp@v0.0.0-20240419123447-f1cffa2c0c49: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/intel-go/cpuid@v0.0.0-20220614022739-219e067757cb: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/kevinburke/ssh_config@v1.2.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/klauspost/compress@v1.17.8: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/pierrec/lz4/v4@v4.1.21: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/rekby/gpt@v0.0.0-20200614112001-7da10aec5566: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/safchain/ethtool@v0.3.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/u-root/gobusybox/src@v0.0.0-20240305154353-d8fbaca23e26: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/u-root/mkuimage@v0.0.0-20240228022452-899a47eaaa31: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/u-root/uio@v0.0.0-20240224005618-d2acac8f3701: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/ulikunitz/xz@v0.5.12: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/crypto@v0.23.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/exp@v0.0.0-20240506185415-9bf2ced13842: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/net@v0.25.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/sys@v0.20.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/term@v0.20.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/text@v0.15.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/tools@v0.21.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        gopkg.in/yaml.v2@v2.4.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        mvdan.cc/sh/v3@v3.8.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        src.elv.sh@v0.20.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/charmbracelet/bubbles@v0.18.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/charmbracelet/bubbletea@v0.26.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/charmbracelet/lipgloss@v0.10.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/cloudflare/circl@v1.3.8: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/erikgeiser/coninput@v0.0.0-20211004153227-1c3628e74d0f: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/jsimonetti/rtnetlink@v1.4.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/mattn/go-runewidth@v0.0.15: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/mdlayher/socket@v0.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/muesli/termenv@v0.15.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/rivo/uniseg@v0.4.7: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/sahilm/fuzzy@v0.1.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/mod@v0.17.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/sync@v0.7.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        gopkg.in/yaml.v3@v3.0.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/ProtonMail/go-crypto@v0.0.0-20221026131551-cf6655e29de4: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/beevik/ntp@v0.3.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/cenkalti/backoff/v4@v4.1.3: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/charmbracelet/bubbles@v0.15.1-0.20230123181021-a6a12c4a31eb: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/charmbracelet/bubbletea@v0.24.1: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/charmbracelet/lipgloss@v0.7.1: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/cloudflare/circl@v1.3.7: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/containerd/console@v1.0.4-0.20230706203907-8f6c4e4faef5: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/gliderlabs/ssh@v0.1.2-0.20181113160402-cbabf5414432: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/gojuno/minimock/v3@v3.0.8: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/google/go-cmp@v0.5.9: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/google/go-tpm@v0.9.1-0.20230914180155-ee6cbcd136f8: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/google/uuid@v1.3.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/hugelgupf/vmtest@v0.0.0-20240228002643-de15f4612e10: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/insomniacslk/dhcp@v0.0.0-20231206064809-8c70d406f6d2: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/intel-go/cpuid@v0.0.0-20200819041909-2aa72927c3e2: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/jsimonetti/rtnetlink@v1.3.5: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/kevinburke/ssh_config@v1.1.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/klauspost/compress@v1.17.4: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/mattn/go-runewidth@v0.0.14: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/mdlayher/socket@v0.5.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/muesli/termenv@v0.15.1: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/pierrec/lz4/v4@v4.1.14: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/rekby/gpt@v0.0.0-20200219180433-a930afbc6edc: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/rivo/uniseg@v0.4.4: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/safchain/ethtool@v0.0.0-20200218184317-f459e2d13664: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/sahilm/fuzzy@v0.1.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/u-root/gobusybox/src@v0.0.0-20240225013946-a274a8d5d83a: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/u-root/mkuimage@v0.0.0-20240225063926-11a3bcc79c2a: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/u-root/uio@v0.0.0-20240209044354-b3d14b93376a: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        github.com/ulikunitz/xz@v0.5.11: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/crypto@v0.21.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/exp@v0.0.0-20240222234643-814bf88cf225: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/mod@v0.15.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/net@v0.23.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/sync@v0.6.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/sys@v0.18.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/term@v0.18.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/text@v0.14.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        golang.org/x/tools@v0.18.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        gopkg.in/yaml.v2@v2.2.8: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        gopkg.in/yaml.v3@v3.0.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        mvdan.cc/sh/v3@v3.7.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
        src.elv.sh@v0.16.0-rc1.0.20220116211855-fda62502ad7f: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod

        To ignore the vendor directory, use -mod=readonly or -mod=mod.
        To sync the vendor directory, run:
                go mod vendor

```

File: cmds/exp/tsort/tsort.go
Comment (jbduncan) [reply cmds/exp/tsort/tsort.go]: I think this is pretty much done, but feel free to un-resolve this thread and challenge me!

**Review by rminnich — COMMENTED** [review]
another question, sorry :-)

Can you make a tsort package and have it implement the interface:
https://medium.com/@kdnotes/sort-sort-interface-in-golang-1d263d96956d

**Comment by rminnich** [comment]
also, I think my interface question was dumb, so nvm. I will test out your PR

File: cmds/exp/tsort/tsort.go
Comment (binjip978) [inline cmds/exp/tsort/tsort.go:57]: Thx for PR!

Could you change order for this arguments to be more consistent with the rest of codebase:
```go
func run(stdin io.Reader, stdout, stderr io.Writer, args []string) error
```
and add something like this in main:
```go
	if err := run(os.Stdin, os.Stdout, os.Stderr, args...); err != nil {
		log.Fatalf("tsort: %v", err)
	}
```

And return an error instead of integer, I think it will be better and maybe you can eliminate os.Stderr if you don't plan to write warnings.

**Comment by rminnich** [comment]
I pulled your PR, removed the vendoring, 
did a go mod tidy
and yeah, it wants to vendor, I'm always a bit mystified on what drives the rules on vendoring.
Thanks.

**Comment by jbduncan** [comment]
> also, I think my interface question was dumb, so nvm. I will test out your PR

@rminnich No worries, I totally understand. I made the same mistake too when first learning about tsort. :grin:

File: cmds/exp/tsort/tsort.go
Comment (jbduncan) [reply cmds/exp/tsort/tsort.go:59]: > Thx for PR!

@binjip978 yw! And thanks for reviewing it. :)

> Could you change order for this arguments to be more consistent with the rest of codebase:
> 
> ```go
> func run(stdin io.Reader, stdout, stderr io.Writer, args []string) error
> ```

Sure! When I next have time, I'll do that.

> and add something like this in main:
> 
> ```go
> 	if err := run(os.Stdin, os.Stdout, os.Stderr, args...); err != nil {
> 		log.Fatalf("tsort: %v", err)
> 	}
> ```
> 
> And return an error instead of integer, I think it will be better and maybe you can eliminate os.Stderr if you don't plan to write warnings.

If we used `log.Fatalf` like you suggest, it would print a timestamp before the message, which existing tsorts don't seem to do. So `2024/05/11 17:15:16 tsort: odd data count` vs `tsort: odd data count`. Would you like me to go ahead with this `log.Fatalf` pattern anyway?

Also, I do plan on writing warnings, because when existing tsorts encounter cycles in the input, they print the cycles to stderr and print the rest of the nodes in topological order to stdout.

Do you have any preferences between these options?

**Option A**
```
a
b
2024/05/11 17:15:16 tsort: cycle in data
2024/05/11 17:15:16 tsort: c
2024/05/11 17:15:16 tsort: d
e
```

**Option B**
```
a
b
2024/05/11 17:15:16 tsort: cycle in data
tsort: c
tsort: d
e
```

**Option C**
```
a
b
tsort: cycle in data
tsort: c
tsort: d
e
```

If option A or B, I'll need to find a way to use `log.Errorf` in `run` or something to make this work.

Comment (binjip978) [reply cmds/exp/tsort/tsort.go:59]: Warnings are fine and you can just write to stderr and continue no logging prefix required. 
But log.Fatal will basically do a print and call os.Exit(1), so I think it can be only the last message. In this case I think it's ok for fatal error with logging message, most of our command doing exactly that. 

But my point mostly to return an error instead of integers, it can be non fatal error, so in main you can do os.Exit(1).

Something like this:
```go
if errors.Is(err, errNonFatal) {
  os.Exit(1)
}
```

Comment (jbduncan) [reply cmds/exp/tsort/tsort.go]: Okay, understood! I've pushed some changes which should address this, please take a look.

**Comment by jbduncan** [comment]
I'm tempted to move the `topologicalOrder` function into `pkg`, but I'm not fond of its callback style, and I'd rather wait until range-over-func is available in Go to migrate the callbacks to an iterator style. So I'll leave `topologicalOrder` where it is for now.

**Comment by jbduncan** [comment]
@rminnich @binjip978 I'm ready for another review when you are!

When we're done, I intend to make two signed-off commits, one for the `github.com/google/go-cmp` upgrade/vendoring and another for `tsort` itself.

**Comment by jbduncan** [comment]
By the way, I noticed that CIFuzz didn't catch a bug I found towards the end of development, so I was wondering if there's a way to tell it to fuzz `tsort` directly, if it doesn't do that already?

**Comment by jbduncan** [comment]
I've squashed many bugs over the last few days, so I'm doubly ready!

File: cmds/exp/tsort/multiset.go
Comment (binjip978) [inline cmds/exp/tsort/multiset.go:9]: Probably newMultiset will be more idiomatic and more consistent with newGraph

File: cmds/exp/tsort/multiset.go
Comment (binjip978) [inline cmds/exp/tsort/multiset.go:13]: I think you can simplify this to
```go
type multiset map[string]int
```

and then you can remove all of m.m

File: cmds/exp/tsort/queue.go
Comment (binjip978) [inline cmds/exp/tsort/queue.go:7]: type queue []string may simplify this as well

Comment (rminnich) [reply cmds/exp/tsort/queue.go:7]: I'm sort of ok either way. If you use a struct, then, later, it is easy to add to it without changing existing code? wdyt?

**Review by rminnich — COMMENTED** [review]
I think we are very close. Let's get this last cycle of reviews done and this can go in.

Thanks for your contribution.

**Comment by rminnich** [comment]
Dont' forget to git commit --amend -s

**Comment by rminnich** [comment]
Also, we like to maintain the rule that the tests work for any commit. Might be best to squash these many commits? Thanks.

File: cmds/exp/tsort/queue.go
Comment (binjip978) [reply cmds/exp/tsort/queue.go:7]: I agree, if there are plans to add something here, sure struct will work better.

File: cmds/exp/tsort/multiset.go
Comment (jbduncan) [reply cmds/exp/tsort/multiset.go:9]: I considered that, but as this function returns a struct directly rather than a pointer (because the struct only contains one field, so making it a pointer sounds redundant to me), I think `make*` is more appropriate than `new*`. WDYT?

File: cmds/exp/tsort/queue.go
Comment (jbduncan) [reply cmds/exp/tsort/queue.go:7]: I was in two minds, as `queue`, `set` and `multiset` are internal data structures. But I imagine they could be used elsewhere in the future, so I'm inclined to leave them be. WDYT?

Comment (jbduncan) [reply cmds/exp/tsort/queue.go:7]: Note: I purposefully made `set` extend `map[string]struct{}`, rather than encapsulate it in a struct, because AFAICT none of the operations that one could do on a map can invalidate a set. For example, all map keys are unique, therefore all set values are unique too. But I'm happy to be challenged on this.

**Comment by jbduncan** [comment]
> I think we are very close. Let's get this last cycle of reviews done and this can go in.
> 
> Thanks for your contribution.

You're very welcome, this was fun to work on!

> Dont' forget to git commit --amend -s

Okay, on it. :+1: 

> Also, we like to maintain the rule that the tests work for any commit. Might be best to squash these many commits? Thanks.

Cool, I was planning to squash and sign the commits sooner or later, I just wasn't sure when you wanted me to do so, so I'm on it.

File: cmds/exp/tsort/multiset.go
Comment (jbduncan) [reply cmds/exp/tsort/multiset.go:13]: See https://github.com/u-root/u-root/pull/2984#discussion_r1605815108

File: cmds/exp/tsort/multiset.go
Comment (binjip978) [reply cmds/exp/tsort/multiset.go:9]: I don't think it matters what kind of stuff this function should return. If you create things like this it more idiomatic to call it new*.

For example from google style guide:
```go
package producer

type Thinger struct{ ... }
func (t Thinger) Thing() bool { ... }

func NewThinger() Thinger { return Thinger{ ... } }
```

File: cmds/exp/tsort/queue.go
Comment (binjip978) [reply cmds/exp/tsort/queue.go:7]: It's fine by me, I don't have any strong preference here. Also I guess if Go standard library will catch up with generics all of this will be gone.

Comment (jbduncan) [reply cmds/exp/tsort/queue.go:7]: Perfect. In that case I'll set this conversation to "resolved". Thanks, both!

File: cmds/exp/tsort/multiset.go
Comment (jbduncan) [reply cmds/exp/tsort/multiset.go]: Ahh okay, thank you for pointing me towards the Google Style Guide on this! I remember learning that `make*` was better from a colleague, but I can't remember why now nor did I ask them why, so I'm more than happy to use `new*` instead.

I've just force-pushed a change, see the latest commit.

**Comment by jbduncan** [comment]
@rminnich @binjip978 I'm ready for another review whenever you are.

**Review by binjip978 — APPROVED** [review]

**Comment by jbduncan** [comment]
I think the code coverage has gone down because, due to the prevalence of maps in this implementation, one or two of the code paths aren't always visited by the tests. I'm not entirely sure what to do about them, so I'm happy to leave it to your judgement.

**Comment by jbduncan** [comment]
@rminnich @binjip978, I see that @binjip978 airlines i things a week ago, so I was wondering if there's anything else I need to do before this can be merged in? 

I'm aware of the drop in code coverage, but I don't think it's an issue. I think it's just flaky because the implementation relies on maps heavily, so one or two of the code paths aren't always visited by the tests.

**Comment by binjip978** [comment]
> I think the code coverage has gone down because, due to the prevalence of maps in this implementation, one or two of the code paths aren't always visited by the tests. I'm not entirely sure what to do about them, so I'm happy to leave it to your judgement.

No worries, I updated the label, but if no new comments will follow I will merge it soon.

**Comment by jbduncan** [comment]
Woo! Really happy to see this merged in, so thank you for reviewing this and being so receptive, @binjip978 and @rminnich. :)

**Pull request merged**
